kedro/io/core.py

```diff
@@ -227,32 +228,23 @@ def save(self, data: _DI) -> None:
             message = f"Failed while saving data to data set {str(self)}.\n{str(exc)}"
             raise DatasetError(message) from exc
 
-    def __str__(self) -> str:
-        def _to_str(obj: Any, is_root: bool = False) -> str:
-            """Returns a string representation where
-            1. The root level (i.e. the Dataset.__init__ arguments) are
-            formatted like Dataset(key=value).
-            2. Dictionaries have the keys alphabetically sorted recursively.
-            3. None values are not shown.
-            """
-
-            fmt = "{}={}" if is_root else "'{}': {}"  # 1
-
-            if isinstance(obj, dict):
-                sorted_dict = sorted(obj.items(), key=lambda pair: str(pair[0]))  # 2
-
-                text = ", ".join(
-                    fmt.format(key, _to_str(value))  # 2
-                    for key, value in sorted_dict
-                    if value is not None  # 3
+    def _pretty_repr(self, object_description: dict[str, Any]) -> str:
+        str_keys = []
+        for arg_name in sorted(object_description, key=lambda key: str(key)):
+            if object_description[arg_name] is not None:
+                descr = pprint.pformat(
+                    object_description[arg_name],
+                    sort_dicts=False,
+                    compact=True,
+                    depth=2,
+                    width=500,
                 )
+                str_keys.append(f"{arg_name}={descr}")
 
-                return text if is_root else "{" + text + "}"  # 1
-
-            # not a dictionary
-            return str(obj)
+        return f"{type(self).__module__}.{type(self).__name__}({', '.join(str_keys)})"
```

Comment:
If it comes from `kedro_datasets`, I feel like it would be much cleaner to have the short name users can provide (i.e. `pandas.CSVDataset` instead of `kedro_datasets.pandas.csv_dataset.CSVDataset`).

Similarly, from Kedro core, `MemoryDataset`, etc. seems much easier to read than the full path.

Reply:
I agree on your points but decided to keep the full path, so `__repr__` is as ambiguous as possible and one could easily understand where to look the implementation of the printed dataset.

Ideally, it would be good to have some `"short=True"` flag for the representation that can be set by user. But by default, I prefer to keep a full module name.

Curious what other think.

Reply:
Happy to hear what others think, but IMO we actually don't use the module name in the docs and examples. The standard way people write entries in catalog is the "short" version,  and I think that should be reflected.

Reply:
We can also consider adding `__str__` with a short representation, so that `__repr__` is actual representation but `__str__` adapted for printing

Reply:
> The standard way people write entries in catalog is the "short" version

Most users we've spoken to create their own datasets sooner or later, so people have a mix between short and long names in the catalog.

Also, the long name would work even for `kedro-datasets` right? The fact that we use the short one is a matter of convention and convenience. But arguably it obscures where the datasets come from, and maybe play a role in people not understanding their tracebacks (especially with `s` vs `S`).

So I'm slightly in favour of keeping the long names for all datasets to avoid introducing an exception to the logic written here, but I don't have a strong opinion.

(In either case, no matter what we do I don't think we should complicate the implementation with a flag.)